File: Stack.rb
class Stack

  def initialize size
    @stack = Array.new size
    @top = -1
    @upper_bound = size
    @lower_bound = 0
  end

  def push num
    if self.size < @upper_bound
      @stack[@top+=1] = num
    else
      raise "out of bounds"
    end
  end

  def pop
    if self.size > @lower_bound
      @stack[@top] = nil
      @top -= 1
    else
      raise "can't pop empty stack"
    end
  end

  def peek
    @stack[@top]
  end

  def size
    @top + 1
  end
end

File: public/javascripts/Stack.js
function Stack(size) {
  this.upperBound = size;
  this.lowerBound = 0;
  this.stack = new Array(size);
  this.top = -1;
}

Stack.prototype.push = function(value) {
  if(!this.isFull()) {
    this.top++;
    this.stack[this.top] = value;
  } else {
    throw new Error('Stack is full');
  }
};

Stack.prototype.pop = function() {
  if(!this.isEmpty()) {
    this.stack[top] = null;
    this.top--;
  } else {
    throw new Error('Stack is empty');
  }
};

Stack.prototype.peek = function() {
  return this.stack[this.top];
};

Stack.prototype.isFull = function() {
  return this.top+1 > this.upperBound-1;
};

Stack.prototype.isEmpty = function() {
  return this.top < 0;
};
